s3/src/main/scala/akka/stream/alpakka/s3/impl/HttpRequests.scala

```diff
@@ -0,0 +1,58 @@
+package akka.stream.alpakka.s3.impl
+
+import scala.concurrent.{ ExecutionContext, Future }
+import akka.http.scaladsl.marshallers.xml.ScalaXmlSupport._
+import akka.http.scaladsl.marshalling.Marshal
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.Uri.Query
+import akka.http.scaladsl.model.headers.Host
+import akka.util.ByteString
+import akka.stream.scaladsl.Source
+import akka.http.scaladsl.model.RequestEntity
+import akka.http.scaladsl.model.RequestEntity
+
+private[alpakka] object HttpRequests {
+
+  def s3Request(s3Location: S3Location, method: HttpMethod = HttpMethods.GET, uriFn: (Uri => Uri) = identity): HttpRequest = {
+    HttpRequest(method)
+      .withHeaders(Host(requestHost(s3Location)))
+      .withUri(uriFn(requestUri(s3Location)))
+  }
+
+  def initiateMultipartUploadRequest(s3Location: S3Location): HttpRequest = {
+    s3Request(s3Location, HttpMethods.POST, _.withQuery(Query("uploads")))
+  }
+
+  def getRequest(s3Location: S3Location): HttpRequest = {
+    s3Request(s3Location)
+  }
+
+  def uploadPartRequest(upload: MultipartUpload, partNumber: Int, payload: Source[ByteString, _], payloadSize: Int): HttpRequest = {
+    s3Request(
+      upload.s3Location,
+      HttpMethods.PUT,
+      _.withQuery(Query("partNumber" -> partNumber.toString, "uploadId" -> upload.uploadId))
+    ).withEntity(HttpEntity(ContentTypes.`application/octet-stream`, payloadSize, payload))
+  }
+
+  def completeMultipartUploadRequest(upload: MultipartUpload, parts: Seq[(Int, String)])(implicit ec: ExecutionContext): Future[HttpRequest] = {
+    val payload = <CompleteMultipartUpload>
+                    {
+                      parts.map { case (partNumber, etag) => <Part><PartNumber>{ partNumber }</PartNumber><ETag>{ etag }</ETag></Part> }
+                    }
+                  </CompleteMultipartUpload>
+    for {
+      entity <- Marshal(payload).to[RequestEntity]
+    } yield {
+      s3Request(
+        upload.s3Location,
+        HttpMethods.POST,
+        _.withQuery(Query("uploadId" -> upload.uploadId))
+      ).withEntity(entity)
+    }
+  }
+
+  def requestHost(s3Location: S3Location): Uri.Host = Uri.Host(s"${s3Location.bucket}.s3.amazonaws.com")
```

Comment:
Hi all, looking forward to the s3 support! I played around with it a bit and noticed a cert related error. The issues seems to be related to having periods "." in bucket names, which breaks the s3 wild card certificate. For example, if I have a bucket "has.dots" then I'll get the following error: Caused by: java.security.cert.CertificateException: No subject alternative DNS name matching has.dots.s3.amazonaws.com found.

I was able to work around the issue by moving the bucket out of the host and placing it in the path (ie not using virtual hosted–style buckets):

```
  def requestHost(s3Location: S3Location): Uri.Host = Uri.Host("s3-us-west-1.amazonaws.com")

  def requestUri(s3Location: S3Location): Uri = Uri(s"/${s3Location.bucket}/${s3Location.key}").withHost(requestHost(s3Location)).withScheme("https")
```

This issue is described in http://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html
Specifically, "When using virtual hosted–style buckets with SSL, the SSL wild card certificate only matches buckets that do not contain periods. To work around this, use HTTP or write your own certificate verification logic."